ext/opentelemetry-ext-wsgi/src/opentelemetry/ext/wsgi/__init__.py

```diff
@@ -0,0 +1,105 @@
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+The opentelemetry-ext-wsgi package provides a WSGI middleware that can be used
+on any WSGI framework (such as Django / Flask) to track requests timing through
+OpenTelemetry.
+"""
+
+import functools
+import wsgiref.util as wsgiref_util
+
+from opentelemetry import trace
+from opentelemetry.ext.wsgi.version import __version__  # noqa
+
+
+class OpenTelemetryMiddleware:
+    """The WSGI application middleware.
+
+    This class is used to create and annotate spans for requests to a WSGI
+    application.
+
+    Args:
+        wsgi: The WSGI application callable.
+    """
+
+    def __init__(
+            self,
+            wsgi,
+            span_context_propagator=None,
+            distributed_context_propagator=None,
+    ):
+        self.wsgi = wsgi
+
+        # TODO: implement context propagation
+        self.span_context_propagator = span_context_propagator
+        self.distributed_context_propagator = distributed_context_propagator
+
+    @staticmethod
+    def _add_request_attributes(span, environ):
+        span.set_attribute("component", "http")
+        span.set_attribute("http.method", environ["REQUEST_METHOD"])
+
+        host = environ.get("HTTP_HOST") or environ["SERVER_NAME"]
+        span.set_attribute("http.host", host)
+
+        url = (
+            environ.get("REQUEST_URI") or
+            environ.get("RAW_URI") or
+            wsgiref_util.request_uri(environ, include_query=False)
+        )
+        span.set_attribute("http.url", url)
+
+    @staticmethod
+    def _add_response_attributes(span, status):
+        status_code, status_text = status.split(" ", 1)
+        span.set_attribute("http.status_text", status_text)
+
+        try:
+            status_code = int(status_code)
+        except ValueError:
+            pass
+        else:
+            span.set_attribute("http.status_code", status_code)
+
+    @classmethod
+    def _create_start_response(cls, span, start_response):
+        @functools.wraps(start_response)
+        def _start_response(status, response_headers, *args):
+            cls._add_response_attributes(span, status)
+            return start_response(status, response_headers, *args)
+
+        return _start_response
+
+    def __call__(self, environ, start_response):
+        """The WSGI application
+
+        Args:
+            environ: A WSGI environment.
+            start_response: The WSGI start_response callable.
+        """
+
+        tracer = trace.tracer()
+        method = environ["REQUEST_METHOD"]
+        path = environ["PATH_INFO"]
+
+        span_name = "[{}]{}".format(method, path)
```

Comment:
What rule do we try to follow? It seems the HTTP server span name is not clearly documented in the [specification](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/data-semantic-conventions.md#http-server). Might need to get the spec clarified.

Reply:
We don't have to block this PR though.

Reply:
I might be missing something, but what isn’t clear in the specs?

> Given an inbound request for a route (e.g. "/users/:userID?" the name attribute of the span SHOULD be set to this route.
If the route can not be determined, the name attribute MUST be set to the RFC 3986 URI path value.

In this case, shouldn’t the span name be the path value?

Reply:
It should follow the client span wording, something like "the name of the span" rather than "the name attribute of the span". `Attribute` is something different https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/api-tracing.md#Set-Attributes.

Reply:
I see where you are going with this and agree with you. Although it looks like misspelling, this need to be clarified.

Reply:
I was following what the opencensus [flask extension](https://github.com/census-instrumentation/opencensus-python/blob/daecc37ea3ec821b36c5a9da4094200f5a01b021/contrib/opencensus-ext-flask/opencensus/ext/flask/flask_middleware.py#L146) was doing. Agree this needs clarification in the spec.